Add CONTRIBUTING guide (#61)
frequenz-floss/frequenz-repo-config-python: cookiecutter/{{cookiecutter.github_repo_name}}/CONTRIBUTING.md

# Contributing to {{cookiecutter.title}}

> 

Reviewer: I'd suggest adding running `protolint` as a recommended step.

Author: OK, the issue with `protolint` is that it is *difficult* to install (not present in Ubuntu basically), but we could point to the `protolint` install instructions and let the user deal with it.

Reviewer: >but we could point to the protolint install instructions and let the user deal with it.

Agreed.

Author: As said in the other comment, I would prefer to first bring `repo-config` on par to the status quo so I can merge this PR and get it to a more usable state. For example I also have a separate issue to add a CI check using `protolint`:

* https://github.com/frequenz-floss/frequenz-repo-config-python/issues/32

Ideally `protolint` should be run by `nox` (but because it is hard to install I was thinking of maybe making it optional). In that case these instructions will get obsoleted soon. Anyway, the `protolint` needs a bit more thought IMHO.

Would it be OK to add a comment to the `protolint` issue saying that when we add it to the CI we should also provide instructions (or an easy way) for the user run it locally?

Reviewer: > Ideally protolint should be run by nox

Why?
I'd not suggest mixing up the protobuf and python toolchains. IMO it will introduce unnecessary coupling between the API specs and the auxiliary python library.

> Would it be OK to add a comment to the protolint issue saying that when we add it to the CI we should also provide instructions (or an easy way) for the user run it locally?

Is it worth waiting for an issue for adding a couple lines?

Author: > I'd not suggest mixing up the protobuf and python toolchains. IMO it will introduce unnecessary coupling between the API specs and the auxiliary python library.

OK, good point, I guess then the API repos are a bit of an alien here, because this `repo-config` is super focused on Python (the repo name is `frequenz-repo-config-python`).

In any case, my long term plan is to completely ditch `nox` (for other reasons). I guess for practical reasons we should still keep a repo type "api" in here, but in some future maybe we want to change that and have this repo only take care of the Python side of the "api" repos.

> Is it worth waiting for an issue for adding a couple lines?

To do it proper, is not just adding a couple of lines, ideally I should test it and it is a distraction as it is not what I wanted to focus on right now, but I guess it will still be quicker to add it rather than keep discussing it here :)

Author: Done.

Reviewer: >but in some future maybe we want to change that and have this repo only take care of the Python side of the "api" repos.

Makes sense.

>To do it proper, is not just adding a couple of lines, ideally I should test it and it is a distraction as it is not what I wanted to focus on right now,

Ok understood. Thanks anyway for adding it.

> ## Build

Reviewer: I'd suggest adding a section labelled `Python` and adding the python related documentation under it.

Author: I think for now it will make the template and instructions more complicated, what about adding it when we add Rust support.

You can use `build` to simply build the source and binary distribution:

```sh
python -m pip install build
python -m build
```

> ## Local development

Reviewer: This is a bit misleading in this scope, because for local development in rust, these instructions are not relevant. As suggested, moving this to a section dedicated to the python library would be clearer.

Author: Replied in https://github.com/frequenz-floss/frequenz-repo-config-python/pull/61#discussion_r1238258821, let's continue the conversation about if we should add a Python section **now** there so we don't duplicate it.

```

Or you can install all development dependencies (`mypy`, `pylint`, `pytest`,
etc.) in one go too:
```sh
python -m pip install -e .[dev]
```

If you don't want to install all the dependencies, you can also use `nox` to
run the tests and other checks creating its own virtual environments:

```sh
python -m pip install .[dev-noxfile]
nox
```

You can also use `nox -R` to reuse the current testing environment to speed up
test at the expense of a higher chance to end up with a dirty test environment.

### Running tests / checks individually

For a better development test cycle you can install the runtime and test
dependencies and run `pytest` manually.

```sh
python -m pip install .[dev-pytest]  # included in .[dev] too

# And for example
pytest tests/test_*.py
```

Or you can use `nox`:

```sh
nox -R -s pytest -- test/test_*.py
```

The same appliest to `pylint` or `mypy` for example:

```sh
nox -R -s pylint -- test/test_*.py
nox -R -s mypy -- test/test_*.py
```

## Releasing

These are the steps to create a new release:

1. Get the latest head you want to create a release from.

2. Update the `RELEASE_NOTES.md` file if it is not complete, up to date, and

>    clean from template comments (`<!-- ... ->`) and empty sections. Submit

Reviewer: ```suggestion
replace template comments (`<!-- ... ->`) and empty sections with `None`. Submit
```

Author: I think we are mixing up styles here. Originally the `RELEASE_NOTES.md` template was thought to completely remove the empty sections, to remove the extra noise of just having and empty section (or as you suggest `None`).

There is a proposal to improve this (https://github.com/frequenz-floss/frequenz-repo-config-python/discussions/52), but I would leave it for a separate PR. The idea of this PR is just to bring `repo-config` on par with the current status quo.

Reviewer: > There is a proposal to improve this (https://github.com/frequenz-floss/frequenz-repo-config-python/discussions/52), but I would leave it for a separate PR. The idea of this PR is just to bring repo-config on par with the current status quo.

Fair enough.
I'd still suggest the following fix: "clean from" => "remove" ("clean from" does not make much sense here)

Author: Done.
